lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -5389,6 +5390,112 @@ ConvertAtenOp<Aten__InterpolateSizeListScaleListOp>::matchAndRewrite(
   return success();
 }
 
+// Template to create support tril support const tensor for aten.tril
+// legalization
+template <typename T>
+Value createTrilConstTensor(PatternRewriter &rewriter, Operation *op,
+                            ArrayRef<int64_t> shape, int64_t h, int64_t w,
+                            int64_t diagonal) {
+  SmallVector<T> vec;
+
+  for (int64_t i = 0; i < h; i++) {
+    for (int64_t j = 0; j < w; j++) {
+      // Positive diagonal value includes as many diagonals above the main
+      // diagonal, while negative diagonal value excludes as many diagonals
+      // below the main diagonal.
+      if (i >= j - diagonal) {
+        vec.push_back(static_cast<T>(1));
+      } else {
+        vec.push_back(static_cast<T>(0));
+      }
+    }
+  }
+
+  return tosa::getConstTensor<T>(rewriter, op, vec, shape).value();
+}
+
+// Function to get tril support const tensor based on input type
+// for aten.tril legalization
+Value getTrilConstTensor(PatternRewriter &rewriter, Operation *op,
+                         ArrayRef<int64_t> shape, int64_t h, int64_t w,
+                         int64_t diagonal, Type type) {
+  return TypeSwitch<Type, Value>(type)
+      .Case<mlir::FloatType>([&](auto) {
+        return createTrilConstTensor<float>(rewriter, op, shape, h, w,
+                                            diagonal);
+      })
+      .Case<mlir::IntegerType>([&](auto intType) {
+        switch (intType.getWidth()) {
+        case 1:
+          return createTrilConstTensor<bool>(rewriter, op, shape, h, w,
+                                             diagonal);
+        case 32:
+          return createTrilConstTensor<int32_t>(rewriter, op, shape, h, w,
+                                                diagonal);
+        case 64:
+          return createTrilConstTensor<int64_t>(rewriter, op, shape, h, w,
+                                                diagonal);
+        }
+        llvm_unreachable("Invalid integer width");
+      });
+}
+
+// Legalization for aten.tril
+template <>
+LogicalResult ConvertAtenOp<AtenTrilOp>::matchAndRewrite(
+    AtenTrilOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  auto self = adaptor.getSelf();
+
+  // Not a ranked tensor type
+  auto selfType = dyn_cast<RankedTensorType>(self.getType());
+  if (!selfType)
+    return rewriter.notifyMatchFailure(
+        op, "Only ranked tensor types are supported");
+
+  // Rank below 2 not accepted
+  auto selfRank = selfType.getRank();
+  if (selfRank <= 1)
+    return rewriter.notifyMatchFailure(
+        op, "Rank 0 and 1 are not accepted as they cause underflow");
+
```

Comment:
Can you please check that the element type is one of the ones accepted by getTrilConstTensor?

Reply:
I have checked and tested passing the element type to getTrilConstTensor, and it's been working correctly so far.